lib/src/message.dart

```diff
@@ -59,16 +58,30 @@ abstract class Message {
   /// Content-Type header, it will be set to "application/octet-stream".
   Message(body,
       {Encoding encoding,
-      Map<String, String> headers,
-      Map<String, Object> context})
-      : this._(new Body(body, encoding), headers, context);
+        Map<String, String> headers,
+        Map<String, Object> context})
+      : this.__(body, _determineMediaType(body, encoding, headers), headers,
+      context);
 
-  Message._(Body body, Map<String, String> headers, Map<String, Object> context)
+  Message.__(body, MediaType contentType, Map<String, String> headers,
+      Map<String, Object> context)
+      : this._(new Body(body, encodingForMediaType(contentType, null)), contentType,
```

Comment:
This `new Body(...)` invocation looks wrong to me... the `Body`'s encoding shouldn't ever come from the headers, it should only come from the explicit encoding passed by the user. So this should continue to be `new Body(body, encoding)`, which means you don't have to do this many-layered-constructor thing anymore.

Reply:
`_determineMediaType` is what determines the encoding for the body. If `encoding` is set directly then it should override what is in `content-type` header so that value can be used when creating the body.

This reflects the behavior of this test https://github.com/dart-lang/http/blob/master/test/request_test.dart#L152-L158

Reply:
The way this is written currently, the encoding passed to `new Body()` can come from `headers`. That's not correct. It should only ever come from the `encoding` parameter, or from the default value (which should be UTF-8).

Reply:
So just to be clear if someone were to set `content-type: 'application/json; charset=iso-8859-1'` and not set `encoding` to LATIN when creating a `Request` it should not encode the body with LATIN it should be UTF8 encoded?

Reply:
Correct, and the header should be updated accordingly.